File: go/jimmy_week1.go
package main

import "fmt"

func ValidParentheses(parens string) bool {
	const (
		OPEN = 40
		CLOSE = 41
	)
	result := 0
	for _, v := range parens {
		if int(v) == OPEN {
			result ++
		} else if int(v) == CLOSE {
			result --
			if result < 0 {
				return false
			}
		}
	}

	return result == 0
}

func main() {
	input1 := "(((())))" // true
	input2 := "(()" //false
	input3 := "()(" // false
	fmt.Println(ValidParentheses(input1))
	fmt.Println(ValidParentheses(input2))
	fmt.Println(ValidParentheses(input3))
}

File: php/jimmy_week1.php
<?php

function ValidParentheses(string $parens)
{
    $result = 0;
    for ($i = 0; $i < strlen($parens); $i++) {
        if ($parens[$i] === "(") {
            $result++;
        }
        if ($parens[$i] === ")") {
            $result--;
        }
        if ($result < 0) {
            return false;
        }
    }
    return $result === 0;
}

var_dump(ValidParentheses("(())"));
var_dump(ValidParentheses("((())"));
var_dump(ValidParentheses(")()"));
var_dump(ValidParentheses("())"));